crates/wasi-common/src/virtfs/pipe.rs

```diff
@@ -0,0 +1,412 @@
+    reader: Rc<RefCell<R>>,
+}
+
+impl<R: Read + Any> ReadPipe<R> {
+    /// Create a new pipe from a `Read` type.
+    ///
+    /// All `Handle` read operations delegate to reading from this underlying reader.
+    pub fn new(r: R) -> Self {
+        Self::from_shared(Rc::new(RefCell::new(r)))
+    }
+
+    /// Create a new pipe from a shareable `Read` type.
+    ///
+    /// All `Handle` read operations delegate to reading from this underlying reader.
+    pub fn from_shared(reader: Rc<RefCell<R>>) -> Self {
+        use types::Rights;
+        Self {
+            rights: Cell::new(HandleRights::new(
+                Rights::FD_DATASYNC
+                    | Rights::FD_FDSTAT_SET_FLAGS
+                    | Rights::FD_READ
+                    | Rights::FD_SYNC
+                    | Rights::FD_FILESTAT_GET
+                    | Rights::POLL_FD_READWRITE,
+                Rights::empty(),
+            )),
+            reader,
+        }
+    }
+
+    /// Try to convert this `ReadPipe<R>` back to the underlying `R` type.
+    ///
+    /// This will fail with `Err(self)` if multiple references to the underlying `R` exist.
+    pub fn try_into_inner(mut self) -> std::result::Result<R, Self> {
+        match Rc::try_unwrap(self.reader) {
+            Ok(rc) => Ok(RefCell::into_inner(rc)),
+            Err(reader) => {
+                self.reader = reader;
+                Err(self)
+            }
+        }
+    }
+}
+
+impl From<Vec<u8>> for ReadPipe<io::Cursor<Vec<u8>>> {
+    fn from(r: Vec<u8>) -> Self {
+        Self::new(io::Cursor::new(r))
+    }
+}
+
+impl From<&[u8]> for ReadPipe<io::Cursor<Vec<u8>>> {
+    fn from(r: &[u8]) -> Self {
+        Self::from(r.to_vec())
+    }
+}
+
+impl From<String> for ReadPipe<io::Cursor<String>> {
+    fn from(r: String) -> Self {
+        Self::new(io::Cursor::new(r))
+    }
+}
+
+impl From<&str> for ReadPipe<io::Cursor<String>> {
+    fn from(r: &str) -> Self {
+        Self::from(r.to_string())
+    }
+}
+
+impl<R: Read + Any> Handle for ReadPipe<R> {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn try_clone(&self) -> io::Result<Box<dyn Handle>> {
+        Ok(Box::new(Self {
+            rights: self.rights.clone(),
+            reader: self.reader.clone(),
+        }))
+    }
+
+    fn get_file_type(&self) -> types::Filetype {
+        types::Filetype::Unknown
+    }
+
+    fn get_rights(&self) -> HandleRights {
+        self.rights.get()
+    }
+
+    fn set_rights(&self, rights: HandleRights) {
+        self.rights.set(rights)
+    }
+
+    fn advise(
+        &self,
+        _advice: types::Advice,
+        _offset: types::Filesize,
+        _len: types::Filesize,
+    ) -> Result<()> {
+        Err(Errno::Spipe)
+    }
+
+    fn allocate(&self, _offset: types::Filesize, _len: types::Filesize) -> Result<()> {
+        Err(Errno::Spipe)
+    }
+
+    fn fdstat_set_flags(&self, _fdflags: types::Fdflags) -> Result<()> {
+        // do nothing for now
```

Comment:
One thing I punted on when first doing virtfs work was figuring out what kind of errors are actually acceptable to return from what function. I broadly worked off of `In the parts of WASI which do correspond to POSIX functionality, WASI follows POSIX when it doesn't conflict with WASI's other goals. And, we consult POSIX even when we aren't strictly following it.` from [DesignPrinciples.md](https://github.com/WebAssembly/WASI/blob/master/docs/DesignPrinciples.md) but we should probably consider documenting the full set of errors that WASI functions may return - even if that's simply "all functions may return any `Errno`" (I think we can and should do better than that, since it would be good to be accurate in communicating kinds of errors to downstream users)

I bring this up mostly because I'm wondering if it's right to accept and do nothing here, or should this `Errno::Perm`? Silently doing nothing is probably the right answer, but there ought to be more implementation guidance for virtfs users.

Reply:
What does POSIX do if you try to set O_APPEND on a pipe? Do real-world implementations support O_NONBLOCK on pipes? I don't have a clear sense of the right thing to do here yet.

I agree that we eventually want to move toward documenting the possible error codes for each function. I expect that'll be easier to do once we can start splitting up "file descriptors" into separate types, and we can have separate errno enums rather than one big POSIX-style errno space shared by filesystem and networking functions of all kinds.

Reply:
Agreed that there should be more implementation guidance, but since I'm new to this corner of the codebase I'm not sure what to provide. Would you be okay with merging this for now with the expectation of doing that work later?

Reply:
> What does POSIX do if you try to set O_APPEND on a pipe? Do real-world implementations support O_NONBLOCK on pipes? I don't have a clear sense of the right thing to do here yet.

I was actually really curious about this and how I'd find the answer: it turns out [Linux source](https://github.com/torvalds/linux/blob/7c30b859a947535f2213277e827d7ac7dcff9c84/fs/fcntl.c#L35-L85) is a pretty clear answer - looks like `O_APPEND` and `O_NONBLOCK` would be accepted without issue. I didn't see anything in particular about `fcntl`'s interaction with pipes in POSIX either, so I assume the flags should do the things they say they do? Where O_APPEND is functionally ignored on pipes, of course :)

Personally merging this as-is and providing better implementation guidance independently is exactly the approach I think we ought to take.

Reply:
Sorry @sunfishcode, Github wasn't showing your comment in this thread when I left my reply. In my testing, `O_APPEND` seems to be ignored.

Reply:
Yes, as long as you're ok making subtle behavior changes later :-).